node/router/src/libp2p_event_handler.rs

```diff
@@ -263,6 +263,31 @@ impl Libp2pEventHandler {
         }
     }
 
+    async fn get_listen_addr_or_add(&self) -> Option<Multiaddr> {
+        if let Some(addr) = self.get_listen_addr() {
+            return Some(addr);
+        }
+
+        let ipv4_addr = public_ip::addr_v4().await?;
+
+        let mut addr = Multiaddr::empty();
+        addr.push(Protocol::Ip4(ipv4_addr));
+        addr.push(Protocol::Tcp(self.network_globals.listen_port_tcp()));
+        addr.push(Protocol::P2p(self.network_globals.local_peer_id().into()));
+
+        self.network_globals
+            .listen_multiaddrs
+            .write()
+            .insert(0, addr.clone());
+
+        info!(
```

Comment:
info? don't think we need to print this out every time we construct the announce file?

Reply:
This will be print out only once.

Note, once public IP detected, it will be added into `listen_addr`.